pocs/camera/camera.py

```diff
@@ -155,14 +172,99 @@ def CCD_cooling_power(self):
+                                                                          headers,
+                                                                          filename,
+                                                                          *args,
+                                                                          **kwargs)
+
+        exposure_event = self.take_exposure(seconds=exp_time, filename=file_path, *args, **kwargs)
+
+        # Add most recent exposure to list
+        if self.is_primary:
+            observation.exposure_list[image_id] = file_path
+
+        # Process the exposure once readout is complete
+        t = Thread(target=self.process_exposure, args=(metadata, camera_event, exposure_event))
+        t.name = '{}Thread'.format(self.name)
+        t.start()
+
+        return camera_event
 
     def take_exposure(self, *args, **kwargs):
         raise NotImplementedError
 
-    def process_exposure(self, *args, **kwargs):
-        raise NotImplementedError
+    def process_exposure(self, info, signal_event, exposure_event=None):
+        """
+        Processes the exposure.
+
+        If the camera is a primary camera, extract the jpeg image and save metadata to mongo
+        `current` collection. Saves metadata to mongo `observations` collection for all images.
+
+        Args:
+            info (dict): Header metadata saved for the image
+            signal_event (threading.Event): An event that is set signifying that the
+                camera is done with this exposure
+            exposure_event (threading.Event, optional): An event that should be set
+                when the exposure is complete, triggering the processing.
+        """
+        # If passed an Event that signals the end of the exposure wait for it to be set
+        if exposure_event is not None:
+            exposure_event.wait()
+
+        image_id = info['image_id']
+        seq_id = info['sequence_id']
+        file_path = info['file_path']
+        self.logger.debug("Processing {} {}".format(image_id))
+
+        try:
+            self.logger.debug("Extracting pretty image")
+            img_utils.make_pretty_image(file_path,
+                                        title=info['field_name'],
+                                        primary=info['is_primary'])
+        except Exception as e:
+            self.logger.warning('Problem with extracting pretty image: {}'.format(e))
+
+        self._process_fits(filepath, info)
+
+        if info['is_primary']:
+            self.logger.debug("Adding current observation to db: {}".format(image_id))
+            self.db.insert_current('observations', info, include_collection=False)
+        else:
+            self.logger.debug('Compressing {}'.format(file_path))
+            fits_utils.fpack(file_path)
+
+        self.logger.debug("Adding image metadata to db: {}".format(image_id))
+        self.db.observations.insert_one({
```

Comment:
We have notes to work on this elsewhere, but if `info` contains items that cannot be serialized (like was happening with `EQUINOX`) then this command fails in which case the `signal_event.set()` never happens and the whole system hangs. We want to clean up that serialization but I would add a `try/except` around this call in the base class.

Reply:
OK.